server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,187 @@
+import org.elasticsearch.index.engine.InternalEngine;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+// TODO(ES-5727): add a retry mechanism to TransportGetFromTranslogAction
+public class TransportGetFromTranslogAction extends HandledTransportAction<
+    TransportGetFromTranslogAction.Request,
+    TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, Request::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final GetRequest getRequest = request.getRequest();
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert getRequest.realtime();
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
+        ActionListener.completeWith(listener, () -> {
+            var result = indexShard.getService()
+                .getFromTranslog(
+                    getRequest.id(),
+                    getRequest.storedFields(),
+                    getRequest.realtime(),
+                    getRequest.version(),
+                    getRequest.versionType(),
+                    getRequest.fetchSourceContext(),
+                    getRequest.isForceSyntheticSource()
+                );
+            long segmentGeneration = -1;
+            if (result == null) {
+                segmentGeneration = ((InternalEngine) engine).getLastUnsafeSegmentGenerationForGets();
+            }
+            return new Response(result, segmentGeneration);
+        });
+    }
+
+    public static class Request extends ActionRequest {
+
+        private final GetRequest getRequest;
+        private final ShardId shardId;
+
+        public Request(GetRequest getRequest, ShardId shardId) {
```

Comment:
`GetRequest` is a `SingleShardRequest` and can contain a `ShardId`, should we check that they are the same?

Reply:
Indeed if it's the same shardId, then we can re-use the `GetRequest` as the request type of the `TransportGetFromTranslogAction`.

Reply:
Thanks for pointing to the comment. I do agree `SingleShardRequest` did not aged well.

If I understand the code correctly, the internal shard id of the GetRequest is resolved when the request is executed through the TransportGetAction/TransportSingleShardAction, and TransportGetAction will be modified for realtime gets on stateless to use the new `TransportGetFromTranslogAction` and the resolved shard id will be passed over the the TransportGetFromTranslogAction.Request.

So we should be fine.